### PR TITLE
SL: Supervised Learning1

### PR DESCRIPTION
**Pull request by steveyuwono**
## Checklists:

- [ ] Read the [CONTRIBUTION](https://github.com/fhswf/MLPro/blob/master/CONTRIBUTING.md) guide (**required**).
- [ ] Update the history on the source code (**required**).

### CODE REVIEW

**Comment by steveyuwono** [comment]
Hi @detlefarend, when you have time later (not urgent), could you review the current progress of the MLPro-SL, and whether it is sufficient for version 1? I have updated the class diagram in the RTD file - appendix 2 of this branch. Additionally, I provided PyTorch-based MLPs in the pool of objects.

If everything is fine, I am going to continue with the following tasks:
- Remove afct_pytorch.py from the pool
- Update MBRL examples and env model pool in MLPro-RL
- Documentation of MLPro-SL in RTD for the 1st version

Thank you!

**Comment by detlefarend** [comment]
Hi @steveyuwono, yep! I’ll review it tonight and let you know...

**Comment by detlefarend** [comment]
HI @steveyuwono, I just reviewed the design. I guess I need to explain my origin draft design. Let's find some time on Friday to discuss it.

**Comment by detlefarend** [comment]
HI @steveyuwono, I just reviewed the design. I guess, I need to explain my origin draft design. Let's find some time on Friday to discuss it.

**Comment by steveyuwono** [comment]
Hi @detlefarend , I have updated MLPro-SL according to our last discussions and would like to ask for your review. Moreover, I have updated the class diagrams of the related updates to make them easier to be analysed. Four main sections need to be reviewed, such as:

  1. MLPro-SL-basics
    - Rename **mlpro.sl.models** to **mlpro.sl.basics**
    - Update the related **__init__** file of the package
    - I added **_hyperparameters_check()** method in this level, thus it can be reused by inheritance sub-packages (e.g. FNN, interpolants, etc.)
    
  2. MLPro-SL-FNN
    - As decided in our last meeting, **forward()** function is located in **FNN** base class
    - Then, I also added **_optimize()** and **_calc_loss()** methods
    - The base **MLP** class is introduced by inheriting **FNN**, where we check all the required hyperparameters to run a general MLP model by defining **_hyperparameters_check()**
    
  3. MLPro-SL-pool-afct-pytorch
    - This provides the PyTorch-related helper functions
    - Is it in the correct directory?
    
  4. MLPro-SL-pool-FNN-pytorch
    - Provide a ready-to-use PyTorch-based MLP network
    
If it makes sense, after releasing the 1st version we need to discuss standardizing SL components (e.g. base classes for activation function, optimizer, layer, etc.) as what we did in RL (Action, Reward, Agent, etc.) as well as discuss **SLScenario** and **SLTraining**. Because I believe dealing with SL components is a little tricky.

Moreover, what is the input data type (**p_dataset**) for offline adaptation in SL? At the moment, I keep them undefined.

Please let me know if something needs to be adjusted or improved. Thank you!

**Comment by detlefarend** [comment]
Hi @steveyuwono, I had a bit time before my next meeting. Here's my feedback...

> Hi @detlefarend , I have updated MLPro-SL according to our last discussions and would like to ask for your review. Moreover, I have updated the class diagrams of the related updates to make them easier to be analysed. Four main sections need to be reviewed, such as:
> 
>     1. MLPro-SL-basics
>        - Rename **mlpro.sl.models** to **mlpro.sl.basics**
>        - Update the related ****init**** file of the package
>        - I added **_hyperparameters_check()** method in this level, thus it can be reused by inheritance sub-packages (e.g. FNN, interpolants, etc.)
> 
Nice!
>     2. MLPro-SL-FNN
>        - As decided in our last meeting, **forward()** function is located in **FNN** base class
>        - Then, I also added **_optimize()** and **_calc_loss()** methods
>        - The base **MLP** class is introduced by inheriting **FNN**, where we check all the required hyperparameters to run a general MLP model by defining **_hyperparameters_check()**
> 
You added it as a sub-package (folder). I think it could be enough to place it as a module fnn.py in folder ./sl and add it to the init-file there... 
>     3. MLPro-SL-pool-afct-pytorch
>        - This provides the PyTorch-related helper functions
>        - Is it in the correct directory?
> 
If they are not related to fnns then it's fine. Otherwise I would place it under fnn
>     4. MLPro-SL-pool-FNN-pytorch
>        - Provide a ready-to-use PyTorch-based MLP network
> 
I propose to place it under ./afct: ./sl/pool/afct/fnn/
> 
> If it makes sense, after releasing the 1st version we need to discuss standardizing SL components (e.g. base classes for activation function, optimizer, layer, etc.) as what we did in RL (Action, Reward, Agent, etc.) as well as discuss **SLScenario** and **SLTraining**. Because I believe dealing with SL components is a little tricky.
> 
> Moreover, what is the input data type (**p_dataset**) for offline adaptation in SL? At the moment, I keep them undefined.
> 
Good question... In streams we have classes StreamProvider and Stream. Here we will probably have something similar: a DataSetProvider and a DataSet...
> Please let me know if something needs to be adjusted or improved. Thank you!
Sorry, timeout. I hope, my feedback helps...

**Comment by steveyuwono** [comment]
> Hi @steveyuwono, I had a bit time before my next meeting. Here's my feedback...
> 
> > Hi @detlefarend , I have updated MLPro-SL according to our last discussions and would like to ask for your review. Moreover, I have updated the class diagrams of the related updates to make them easier to be analysed. Four main sections need to be reviewed, such as:
> > ```
> > 1. MLPro-SL-basics
> >    - Rename **mlpro.sl.models** to **mlpro.sl.basics**
> >    - Update the related ****init**** file of the package
> >    - I added **_hyperparameters_check()** method in this level, thus it can be reused by inheritance sub-packages (e.g. FNN, interpolants, etc.)
> > ```
> 
> Nice!
> 
> > ```
> > 2. MLPro-SL-FNN
> >    - As decided in our last meeting, **forward()** function is located in **FNN** base class
> >    - Then, I also added **_optimize()** and **_calc_loss()** methods
> >    - The base **MLP** class is introduced by inheriting **FNN**, where we check all the required hyperparameters to run a general MLP model by defining **_hyperparameters_check()**
> > ```
> 
> You added it as a sub-package (folder). I think it could be enough to place it as a module fnn.py in folder ./sl and add it to the init-file there...
> 
> > ```
> > 3. MLPro-SL-pool-afct-pytorch
> >    - This provides the PyTorch-related helper functions
> >    - Is it in the correct directory?
> > ```
> 
> If they are not related to fnns then it's fine. Otherwise I would place it under fnn
> 
> > ```
> > 4. MLPro-SL-pool-FNN-pytorch
> >    - Provide a ready-to-use PyTorch-based MLP network
> > ```
> 
> I propose to place it under ./afct: ./sl/pool/afct/fnn/
> 
> > If it makes sense, after releasing the 1st version we need to discuss standardizing SL components (e.g. base classes for activation function, optimizer, layer, etc.) as what we did in RL (Action, Reward, Agent, etc.) as well as discuss **SLScenario** and **SLTraining**. Because I believe dealing with SL components is a little tricky.
> > Moreover, what is the input data type (**p_dataset**) for offline adaptation in SL? At the moment, I keep them undefined.
> 
> Good question... In streams we have classes StreamProvider and Stream. Here we will probably have something similar: a DataSetProvider and a DataSet...
> 
> > Please let me know if something needs to be adjusted or improved. Thank you!
> 
> Sorry, timeout. I hope, my feedback helps...

Alright, thank you for the quick response and suggestion. I will update them accordingly

**Comment by steveyuwono** [comment]
Hi @detlefarend , I have updated MLPro-SL according to our last meeting, including removing envmodels from the pool of objects and incorporating them and MLPro-SL to all MBRL howtos. I think this should be enough for version 1 and the next improvements are going to be made after version 1 is released.

If it is fine for you, I will merge this to the main branch and then update the SL-related documentation in the quality assurance branch tomorrow. Thank you!

**Comment by detlefarend** [comment]
> Hi @detlefarend , I have updated MLPro-SL according to our last meeting, including removing envmodels from the pool of objects and incorporating them and MLPro-SL to all MBRL howtos. I think this should be enough for version 1 and the next improvements are going to be made after version 1 is released.
> 
> If it is fine for you, I will merge this to the main branch and then update the SL-related documentation in the quality assurance branch tomorrow. Thank you!

Hi @steveyuwono, I just checked the current state. The SL code itself looks good and as discussed. I have just some doubts regarding the mb howtos. Both implement specific EnvModel child classes. But this is not intended and should not be how we propose/demonstrate. We may have a misunderstanding. EnvModel is not a template class but is ready to use. I don't know whether it performs. I suggest merging the branch to quality assurance, and I quickly implement a model-based derivate of howto_rl_agent_011_train_and_reload_single_agent_gym.py. I am not sure whether it performs. Let's see...

**Comment by detlefarend** [comment]
@steveyuwono there are conflicts to resolve regarding files that were also changed on the qa branch.

**Comment by steveyuwono** [comment]
> @steveyuwono there are conflicts to resolve regarding files that were also changed on the qa branch.

It was conflicting in the unit test and also the bf part. But now it should be solved.

**Comment by detlefarend** [comment]
Hi @steveyuwono, I added a new draft **howto_rl_mb_001_train_and_reload_model_based_agent_gym.py** in branch qa. It shall show the regular way to setup an env model. In line 76 the specific hyperparameters can be added. 



Finally, these hp are handed over to the custom method PyTorchMLP._init_hyperparam(). Here, the hyperparameter space and tuple is to be initialized. See also: class Model, constructor and _init_hyperparam().



I know, it's tricky. That's why we need to explain and demonstrate it. We can have a zoom tomorrow, if you like.

**Review by detlefarend — APPROVED** [review]
Hi Steve, everything looks fine and as discussed. I propose to merge to main and finalize the howto(s) for model-based agents in branch qa...

**Comment by steveyuwono** [comment]
> Hi @steveyuwono, I added a new draft **howto_rl_mb_001_train_and_reload_model_based_agent_gym.py** in branch qa. It shall show the regular way to setup an env model. In line 76 the specific hyperparameters can be added.
> 
> 
> 
> Finally, these hp are handed over to the custom method PyTorchMLP._init_hyperparam(). Here, the hyperparameter space and tuple is to be initialized. See also: class Model, constructor and _init_hyperparam().
> 
> 
> 
> I know, it's tricky. That's why we need to explain and demonstrate it. We can have a zoom tomorrow, if you like.

I think I understand what you mean. I will update the SL-related tasks in qa branch. Thank you
